### PR TITLE
Intro to DAML Chapter 9

### PR DESCRIPTION
**Pull request by bame-da**
CHANGELOG_BEGIN
CHANGELOG_END

Splitting up #7277

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/daml/intro/9_Functional101.rst [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+- `Learn You a Haskell for Great Good! (Miran Lipovača) <http://learnyouahaskell.com/>`__
+- `Programming in Haskell (Graham Hutton) <http://www.cs.nott.ac.uk/~pszgmh/pih.html>`__
+- `Real World Haskell (Bryan O'Sullivan, Don Stewart, John Goerzen) <http://book.realworldhaskell.org/>`__
+
+When comparing DAML to Haskell it's worth noting:
+
+-   Haskell is a lazy language, which allows you to write things like ``head [1..]``, meaning "take the first element of an infinite list". DAML by contrast is strict. Expressions are fully evaluated, which means it it not possible to work with infinite data structures.
+- DAML has a ``with`` syntax for records, and dot syntax for record field access, neither of which present in Haskell. But DAML supports Haskell's curly brace record notation.
+- DAML has a number of Haskell compiler extensions active by default.
+- DAML doesn't support all features of Haskell's type system. For example, there are no existential types of GADTs.
```
Comment (cocreature) [inline docs/source/daml/intro/9_Functional101.rst:36]:
```suggestion
- DAML doesn't support all features of Haskell's type system. For example, there are no existential types or GADTs.
```

File: docs/source/daml/intro/9_Functional101.rst [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+- Actions, called Monads in Haskell.
```
Comment (cocreature) [inline docs/source/daml/intro/9_Functional101.rst:37]:
```suggestion
- Actions are called Monads in Haskell.
```

File: docs/source/daml/intro/9_Functional101.rst [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+  :end-before: -- VOID_END
+
+``void`` also helps express control flow of the type "Create a ``T`` only if a condition is met.
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- CUSTOM_WHEN_BEGIN
+  :end-before: -- CUSTOM_WHEN_END
+
+Note that we still need the ``else`` clause of the same type ``()``. This pattern is so common, it's encapsulated in the standard library function ``DA.Action.when : (Applicative f) => Bool -> f () -> f ()``.
```
Comment (cocreature) [inline docs/source/daml/intro/9_Functional101.rst:333]: Maybe worth mentioning that `when` is also control flow in the sense that it short circuits.

File: docs/source/daml/intro/9_Functional101.rst [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+
+Looping
+.......
+
+Other than branching, the most common form of control flow is looping. Looping is usually used to iteratively modify some state. We'll use pseudocode in this section to illustrate the procedural way of doing things. 
+
+.. code-block:: JavaScript
+
+  function sum(intArr) {
+    int result = 0;
```
Comment (cocreature) [inline docs/source/daml/intro/9_Functional101.rst:364]: Those examples seem like a somewhat weird mix of javascript and not quite javascript. Maybe just go for full javascript?

File: docs/source/daml/intro/9_Functional101.rst [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+  :language: daml
+  :start-after: -- UPGRADE_CONTRACTS_BEGIN
+  :end-before: -- UPGRADE_CONTRACTS_END
+
+Here we have a list ``assetCids : [ContractId AssetV1.Asset]`` and a function ``upgradeAsset : ContractId AssetV1.Asset -> Update (ContractId AssetV2.Asset)``. We want ``UpgradeContracts`` to return a list ``[ContractId AssetV1.Asset]``. Using the map function almost gets us there. ``map upgradeAsset assetCids : [Update (ContractId AssetV2.Asset)]``. This is a list of actions, each resulting in a ``ContractId AssetV2.Asset``. What we need is an ``Update`` action resulting in a ``[ContractId AssetV2.Asset]``. The list and ``Update`` are the wrong way around for our purposes.
+
+
+ Intuitively, it's clear how to fix this: we want the compound action consisting of performing each of the actions in the list in turn. There's a function for that, of course. ``sequence : : Applicative m => [m a] -> m [a]`` implements that intuition and allows us to take the ``Update`` out of the list. So we could write ``sequence (map upgradeAsset assetCids``. This is so common that it's encapsulated in the ``mapA`` function, a possible implementation of which is
```
Comment (cocreature) [inline docs/source/daml/intro/9_Functional101.rst:475]:
```suggestion
 Intuitively, it's clear how to fix this: we want the compound action consisting of performing each of the actions in the list in turn. There's a function for that, of course. ``sequence : : Applicative m => [m a] -> m [a]`` implements that intuition and allows us to take the ``Update`` out of the list. So we could write ``sequence (map upgradeAsset assetCids)``. This is so common that it's encapsulated in the ``mapA`` function, a possible implementation of which is
```
Maybe worth pointing out that `mapA` is more efficient.

**Review by cocreature — APPROVED** [review]
nice, thank you!

File: docs/source/daml/intro/9_Functional101.rst (hunk 3 above)
Comment (bame-da) [reply docs/source/daml/intro/9_Functional101.rst:333]: It only short circuits Action execution, though, not evaluation.

File: docs/source/daml/intro/9_Functional101.rst (hunk 4 above)
Comment (bame-da) [reply docs/source/daml/intro/9_Functional101.rst:364]: But then I'd have to test that it's valid javascript 😢

File: docs/source/daml/intro/9_Functional101.rst (hunk 3 above)
Comment (cocreature) [reply docs/source/daml/intro/9_Functional101.rst:333]: it does short circuit evaluation. It’s hacked up in the compiler

File: docs/source/daml/intro/9_Functional101.rst (hunk 4 above)
Comment (bame-da) [reply docs/source/daml/intro/9_Functional101.rst]: Ok, that was easier than I thought. It's now actual JavaScript

File: docs/source/daml/intro/9_Functional101.rst (hunk 3 above)
Comment (bame-da) [reply docs/source/daml/intro/9_Functional101.rst:333]: I had no idea!

**Pull request merged**